### PR TITLE
python3Packages.icontract: init at 2.6.1

### PR DESCRIPTION
**Pull request by gador**
###### Description of changes

Add [icontract](https://github.com/Parquery/icontract) a design-by-contract library to python.

Also includes dpcontracts, which is needed for icontract.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/dpcontracts/default.nix [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{ lib
+, buildPythonPackage
+, pythonOlder
+, fetchFromGitHub
+}:
+
+buildPythonPackage rec {
+  pname = "dpcontracts";
+  # does not have any kind of versioning
+  version = "45cb8542272c2ebe095c6efb97aa9407ddc8bf3c";
```
Comment (dotlambda) [inline pkgs/development/python-modules/dpcontracts/default.nix:10]:
```suggestion
  version = "unstable-YYYY-MM-DD";
```
You could ask upstream to cut a release.

File: pkgs/development/python-modules/dpcontracts/default.nix [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  disabled = pythonOlder "3.5";
+
+  src = fetchFromGitHub {
+    owner = "deadpixi";
+    repo = "contracts";
+    rev = version;
+    hash = "sha256-FygJPXo7lZ9tlfqY6KmPJ3PLIilMGLBr3013uj9hCEs=";
+  };
+
+  # package does not have any tests
```
Comment (dotlambda) [inline pkgs/development/python-modules/dpcontracts/default.nix:21]:
```suggestion
  # package does not have any tests
  doCheck = false;
```

**Review by dotlambda — CHANGES_REQUESTED** [review]

**Comment by gador** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 173140` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
4 packages built:
</summary>
<ul>
<li>python310Packages.dpcontracts</li>
<li>python310Packages.icontract</li>
<li>python39Packages.dpcontracts</li>
<li>python39Packages.icontract</li>
</ul>
</details>

**Comment by dotlambda** [comment]
Please squash your changes into the appropriate commit.

File: pkgs/development/python-modules/icontract/default.nix [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  propagatedBuildInputs = [
+    asttokens
+    typing-extensions
+  ];
+
+  checkInputs = [
+    pytestCheckHook
+
+    mypy
+    pylint
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/icontract/default.nix:58]:
```suggestion
```

File: pkgs/development/python-modules/icontract/default.nix [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    yapf
+    tox
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/icontract/default.nix:60]:
```suggestion
```
we don't want to use any venvs.

File: pkgs/development/python-modules/icontract/default.nix [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    pydocstyle
+    coverage
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/icontract/default.nix:62]:
```suggestion
```

File: pkgs/development/python-modules/icontract/default.nix [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+buildPythonPackage rec {
+  pname = "icontract";
+  version = "2.6.1";
+  format = "setuptools";
+  disabled = pythonOlder "3.6";
+
+  src = fetchFromGitHub {
+    owner = "Parquery";
+    repo = pname;
+    rev = "refs/tags/v${version}";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/icontract/default.nix:38]:
```suggestion
    rev = "v${version}";
```

File: pkgs/development/python-modules/icontract/default.nix [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    hash = "sha256-QyuegyjVyRLQS0DjBJXpTDNeBM7LigGJ5cztVOO7e3Y=";
+  };
+
+  preCheck = ''
+    # needs to be set for tests
+    # see https://github.com/Parquery/icontract/blob/aaeb1b06780a34b05743377e4cb2458780e808d3/precommit.py#L57
+    export ICONTRACT_SLOW=1
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/icontract/default.nix:45]: This is an assignment. 

We also do not want to run things through extra wrappers like precommit which enforce style checking which we are not interested in.

File: pkgs/development/python-modules/icontract/default.nix (hunk 6 above)
Comment (gador) [reply pkgs/development/python-modules/icontract/default.nix:38]: This is intentional. See [here](https://github.com/NixOS/nixpkgs/commit/7f5915b04801f383a3504a011601e2fb09dfedf8)

File: pkgs/development/python-modules/icontract/default.nix (hunk 7 above)
Comment (gador) [reply pkgs/development/python-modules/icontract/default.nix:45]: we don't. This is why I explicitly exported `ICONTRACT_SLOW` and do not run the pre-commit script

Comment (SuperSandro2000) [reply pkgs/development/python-modules/icontract/default.nix:36]: Can you update the comment to better reflect that? I don't think that is very clear.

File: pkgs/development/python-modules/icontract/default.nix (hunk 6 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/icontract/default.nix:29]: That was a bug in the updater. The repo has no branch named like the tag.

Comment (gador) [reply pkgs/development/python-modules/icontract/default.nix:29]: but why not use the `refs/tags` prefix? This way it is more explicit in the (unlikely) event a branch will be named like the tag in the future

File: pkgs/development/python-modules/icontract/default.nix (hunk 7 above)
Comment (gador) [reply pkgs/development/python-modules/icontract/default.nix:37]: done

**Pull request merged**